src/python/pants/core/goals/clean.py

```diff
@@ -0,0 +1,158 @@
+from typing import Union
+
+import humanize
+
+from pants.base.build_environment import get_pants_cachedir
+from pants.core.util_rules.distdir import DistDir
+from pants.engine.console import Console
+from pants.engine.goal import Goal, GoalSubsystem
+from pants.engine.internals.options_parsing import _Options
+from pants.engine.rules import collect_rules, goal_rule
+from pants.option.global_options import LocalStoreOptions
+from pants.util.dirutil import rm_rf
+
+# N.B. We use a list so that named-caches can go last, as it is quite large and for dry-run
+# it's a nicer UX if it's last.
+_NAMED_CACHE_FLAGS_INFO = [
+    (
+        "process-execution-temp-dirs",
+        "Whether to keep the process execution temporary directories. These are normally cleaned up"
+        " automatically, however they are leaked if --no-process-execution-local-cleanup.",
+    ),
+    (
+        "lmdb-store",
+        "Whether to keep the local file store, which stores the results of subprocesses run by Pants.",
+    ),
+    (
+        "run-tracker",
+        "Whether to clean directories persisted after using `./pants run`.",
+    ),
+    (
+        "repl-temp-dirs",
+        "Whether to clean directories persisted after using `./pants repl`.",
+    ),
+    ("dist-dir", "Whether to clean the distributable files directory. "),
+    ("bootstrap-dir", "Whether to clean directories used by pants for bootstrapping itself."),
+    ("named-caches", "Whether to clean the named global caches."),
+]
+
+
+def _dir_info(*paths: Path):
+    total_size = num_files = 0
+    for path in paths:
+        for dirpath, _, filenames in os.walk(path):
+            total_size += sum(os.path.getsize(os.path.join(dirpath, f)) for f in filenames)
+            num_files += len(filenames)
+    return total_size, num_files
+
+
+class CleanSubsystem(GoalSubsystem):
+    name = "clean"
+    help = "Clean pants caches and temp dirs (based on current settings)"
+
+    @classmethod
+    def register_options(cls, register) -> None:
+        super().register_options(register)
+        register(
+            "--dry-run",
+            advanced=False,
+            type=bool,
+            default=True,
+            help=("Whether to just report what would be cleaned. @TODO: more"),
+        )
+        register(
+            "--all",
+            advanced=False,
+            type=bool,
+            default=False,  # @TODO: True?
+            help=(
+                "Clear all caches. Can be combined with the advanced options for further control."
+            ),
+        )
+
+        # @TODO: Is there a way to collect these programmatically?
+        for flag_slug, help in _NAMED_CACHE_FLAGS_INFO:
+            register(
+                f"--keep-{flag_slug}",
+                advanced=True,
+                type=bool,
+                default=True,
+                help=help,
+            )
+
+
+class Clean(Goal):
+    subsystem_cls = CleanSubsystem
+
+
+@goal_rule
+async def clean(
```

Comment:
For the people who have seen into the other languages' crystal ball. Are there other directories that might need cleaning that would necessitate this to also have a Union Membership?

Reply:
There shouldn't be... the goal of the `named_caches` directory is to be the centralized dumping ground for caches used by all languages. If `@rules` are putting caches anywhere else, there is a very high probability that it is a bug.